File: backend/app/services/symbols.py
import re


_A_SHARE_SYMBOL_PATTERN = re.compile(
    r"^(?:(?P<prefix>SH|SZ|BJ)[.\-_\s]*)?(?P<code>\d{6})(?:[.\-_\s]*(?P<suffix>SH|SZ|BJ))?$",
    re.IGNORECASE,
)


def normalize_a_share_symbol(value: str | None) -> str:
    text = str(value or "").strip().upper()
    compact = re.sub(r"\s+", "", text)
    match = _A_SHARE_SYMBOL_PATTERN.match(compact)
    if match:
        return match.group("code")
    return text


def looks_like_a_share_symbol(value: str | None) -> bool:
    symbol = normalize_a_share_symbol(value)
    return len(symbol) == 6 and symbol.isdigit()

File: frontend/src/lib/symbols.ts
export function normalizeAShareSymbol(value: unknown) {
  const text = String(value ?? '').trim().toUpperCase()
  const compact = text.replace(/\s+/g, '')
  const match = compact.match(/^(?:(SH|SZ|BJ)[.\-_\s]*)?(\d{6})(?:[.\-_\s]*(SH|SZ|BJ))?$/i)
  return match ? match[2] : text
}
